src/coreclr/jit/emit.h

```diff
@@ -2288,31 +2292,134 @@ class emitter
         return (emitCurIG && emitCurIGfreeNext > emitCurIGfreeBase);
     }
 
+#define EMIT_MAX_IG_INS_COUNT 256
+
+#if EMIT_BACKWARDS_NAVIGATION
+#define EMIT_MAX_PEEPHOLE_INS_COUNT 32 // The max number of previous instructions to navigate through for peepholes.
+#endif                                 // EMIT_BACKWARDS_NAVIGATION
+
     instrDesc* emitLastIns;
     insGroup*  emitLastInsIG;
 
 #if EMIT_BACKWARDS_NAVIGATION
     unsigned emitLastInsFullSize;
 #endif // EMIT_BACKWARDS_NAVIGATION
 
+    // Check to see if the last instruction is available.
+    inline bool emitHasLastIns() const
+    {
+        return (emitLastIns != nullptr);
+    }
+
+    // Checks to see if we can cross between the two given IG boundaries.
```

Comment:
This function is very confusing. It would benefit from a full, normal function header.

In general, here and elsewhere in your change, you use the term `prevIG` which is confusing, because you use it to mean "the IG of the previous instruction", not "the previous IG in the IG list". I think you should rename it to be clear, e.g., `prevInsIG`. And, to be clear, here you could have `curInsIG` instead of just `ig`.

Reply:
Another thing is, it is called even if there is no "IG boundary".

Maybe name it `isInsIGSafeForPeepholeOptimization`?

Reply:
Sure, I will try to refactor some of this.